File: hin/_add.py
"""
hin._add
========
"""
from pathlib import Path as _Path

from rich.console import Console as _Console

from . import _decorators
from ._actions import Move as _Move
from ._config import Config as _Config
from ._file import Entry as _Entry
from ._gitignore import Gitignore as _Gitignore
from ._gitignore import unignore as _unignore


def _add_dir(out: _Console, entry: _Entry, config: _Config) -> None:
    gitignore = _Gitignore(entry.value.path)
    gitignore.make()
    for existing in dict(config).values():
        if existing.is_child_of(entry) and isinstance(existing, _Entry):
            # move the child back under the new dotfile and remove
            # from config
            child = entry.child(existing)
            _Move(child).do()
            config.remove(existing)
            out.print(f"unlinked {existing.key.path}")
            out.print(f"committing {existing.key.path}")
            # ensure the child remains versioned by including it in
            # all .gitignore files back up to the parent
            _unignore(child.value.path, entry.value.path)


def _add_child(entry: _Entry, config: _Config) -> bool:
    for existing in dict(config).values():
        if entry.is_child_of(existing):
            # exclude this file from .gitignore
            child = existing.child(entry)
            _unignore(child.value.path, child.value.path.parent)
            return True

    return False


@_decorators.console
@_decorators.config
@_decorators.repository
def add(config: _Config, out: _Console, file: str) -> str:
    """Add new FILE to version control.

    If the file is a symlink not related to the dotfile repo
    then the source of the symlink will be added. The original path will
    be symlinked to the versioned file.

    If the source file is a child of an already added directory it will
    be added as part of the directory. If the source file is a child of
    a directory that has not been added it will be installed
    individually, with the parents recreated, if they do not already
    exist.

    If the source file is a directory then only the directory will be
    added, unless there are child files that have already been added, in
    which case they will be unlinked and added to version control under
    the source.

    Changes will be committed.

    \f

    :param config: Config object.
    :param out: Console object.
    :param file: Dotfile to add.
    :return: Git commit message.
    """
    entry = _Entry.new(file)
    if str(entry.key) in config:
        raise TypeError(f"{entry.key.path} already added")

    if entry.key.path.is_symlink():
        # the target is now the file that the symlink is pointing to
        entry = entry.linksrc

    if entry.value.path.exists():
        # ensure no overwriting of files
        entry = entry.timestamped

    try:
        _Move(entry).do()
        if entry.value.path.is_dir():
            # add .gitignore files and check for any children that might
            # already be versioned
            _add_dir(out, entry, config)

        config.add(entry)
    except FileNotFoundError as err:
        # if move fails then check whether a child can be resolved to a
        # link by checking the dotfile repository internally
        # if selection is a child of an existing dotfile then exclude it
        # from the .gitignore files and commit it
        if not _add_child(entry, config):
            # the selection simply does not exist
            raise err

    out.print(f"added {entry.key.path}")
    return f"add {entry.key.path.relative_to(_Path.home())}"

File: src/add.rs
use std::{fs, path::PathBuf};

use color_eyre::Result;
use log::debug;

use crate::{
    config::Config,
    files::{FileTrait, Matrix},
    git::Git,
    gitignore::{unignore, Gitignore},
};

pub fn add(file: String, mut config: Config, git: Git) -> Result<()> {
    let mut entry = Matrix::new(
        &file,
        &PathBuf::from(&file)
            .file_name()
            .unwrap()
            .to_os_string()
            .into_string()
            .unwrap(),
    );
    for (_, prop) in &config.ini {
        debug!("checking if config contains {}", &entry.key.repr());
        if prop.contains_key(&entry.key.repr()) {
            panic!("{} already added", &entry.key.repr())
        }
    }
    debug!("config does not contain {}", &entry.key.repr());
    if entry.key.path().is_symlink() {
        entry = entry.linksrc()?
    }
    if entry.value.path().exists() {
        entry = entry.timestamped()
    }
    debug!(
        "moving {} to {}",
        &entry.key.path().display(),
        &entry.value.path().display()
    );
    let mut remove_from = false;
    match fs::rename(&entry.key.path(), &entry.value.path()) {
        Ok(_) => {
            if entry.key.path().is_dir() {
                // add .gitignore files and check for any children that
                // might already be versioned
                let gitignore = Gitignore::new(&entry.value.path());
                gitignore.make();
                for (_, prop) in &config.ini {
                    for (key, value) in prop.iter() {
                        let existing =
                            Matrix::new(&key.to_string(), &value.to_string());
                        if existing.is_child_of(&entry)
                            && existing.value.repr().starts_with("$DOTFILES")
                        {
                            // move the child back under the new dotfile
                            // and remove from config
                            let child = entry.child(&existing);
                            fs::rename(child.key.path(), child.value.path())
                                .unwrap();
                            remove_from = true;
                            println!(
                                "unlinked {}",
                                existing.key.path().display()
                            );
                            println!(
                                "committing {}",
                                existing.key.path().display()
                            );
                            // ensure the child remains versioned by
                            // including it in all .gitignore files back
                            // up to the parent
                            unignore(&child.value.path(), &entry.value.path());
                        }
                    }
                }
            }
            if remove_from {
                config.remove(
                    &entry.key.path().into_os_string().into_string().unwrap(),
                );
            }
            config.add(&entry.key.repr(), &entry.value.repr());
            debug!(
                "added {}: {} to config",
                &entry.key.repr(),
                &entry.value.repr()
            );
        }
        Err(e) => {
            let mut add_child = false;
            for (_, prop) in &config.ini {
                for (key, value) in prop.iter() {
                    let existing =
                        Matrix::new(&String::from(key), &String::from(value));
                    if entry.is_child_of(&existing) {
                        let child = existing.child(&entry);
                        unignore(&child.value.path(), &entry.value.path());
                        add_child = true;
                        break;
                    }
                }
            }
            if !add_child {
                panic!("{}", e)
            }
        }
    }
    git.add(vec![
        "dotfiles.ini".to_string(),
        entry.value.relpath().into_os_string().into_string().unwrap(),
    ])?;
    git.commit(&format!("add {}", entry.key.relpath().display()))?;
    println!("added {}", &entry.key.path().display());
    Ok(())
}
